solr/modules/aws-secret-provider/README.md

```diff
@@ -0,0 +1,80 @@
+Apache Solr - AWS Secret Provider
+===========================
+
+An implementation of `SecretCredentialsProvider` that pulls Zookeeper credentials from an AWS Secret Manager.
+
+This plugin uses the [default AWS credentials provider chain](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html), so ensure that your credentials are set appropriately (e.g., via env var, or in `~/.aws/credentials`, etc.).
+
+## Usage
+
+- To enable this feature copy the jar files in `modules/aws-secret-provider/lib` to `SOLR_INSTALL/server/solr-webapp/webapp/WEB-INF/lib/` and add follow the below steps before restarting Solr.
```

Comment:
Thank you very much @gerlowskija for your reviews and time on this.

Yeah this is where I was stuck for several days (weeks)…. The issue here is that classes in _contrib/module_ are not available to _Solrj_  (throwing _ClassNotFoundException_), except if I copied the jars to _solr-webapp/webapp/WEB-INF/lib_.

_SolrZkClient_ (solrj module) is where ZK ACL related classes are instantiated. If the class you’re trying to instantiate is part of a _contrib/module_ you end up with _ClassNotFoundException_ error.


Using the lib dir option doesn’t help because:
1) you would need to add the jars paths into _solrconfig.xml_, but this is a cluster wide feature (not specific to some collection)... 
2) more importantly, Solrj needs first to connect to ZK before it reads _solrconfig_... So the lib needs to be ready before then... it has to be available at the start up before loading anything form zookeeper.

In nutshell, the lib needs to be available to _SolrZkClient_ (= solrj module)

Will take a look to this new [SOLR_MODULES env-var support](https://github.com/apache/solr/blob/main/solr/solr-ref-guide/modules/configuration-guide/pages/solr-modules.adoc) feature (it wasn't there when I started this).

Reply:
I spent the morning trying out the `SOLR_MODULES` and `<sharedLib>` setups and can confirm that I also hit ClassNotFoundExceptions on Solr startup.

Both of those config options (`SOLR_MODULES` and `<sharedLib>`) work by adding the lib directory for the contrib/module to the classloader used by Solr's "CoreContainer" class.  From there it's the solr server's job to use this particular classloader when instantiating any pluggable classes.

We are doing this correctly for at least some of the relevant instantiations (e.g. the ZkCredentialsInjector and ZkCredentialsProvider instantiations [here](https://github.com/apache/solr/pull/826/files#diff-5b63503605ede4384429e74d1fa0c410adc5da8f3246e8c36e49feff2f3ea692R333)).  But we _aren't_ doing this when instantiating "SecretCredentialsProvider": in that case [the instantiating code in SecretCredentialInjector uses the default classloader](https://github.com/apache/solr/pull/826/files#diff-a01f74c25f0b6595287532fe1e307d201c32adeed1bb9e9eb03a38115887449dR77) so it can't "see" AWSSecretCredentialsProvider.

The way I see it, we've got a couple options:

1. We could add a setter to the "ZkCredentialInjector" interface to provide injectors with a classloader that they can use to load any pluggable classes they might need.  The Solr server code could invoke this setter to pass the CoreContainer loader down to the SolrJ code.  This strikes me as a little kludgey, but I don't have any particularly good reason not to like it.
2. Alternatively: the server code in ZkController already uses the correct classloader when it instantiates ZkCredentialInjector instances.  If we made a ZkCredentialInjector that was AWS specific (in effect merging AWSSecretCredentialProvider and SecretCredentialInjector), then the loading problem would disappear entirely.  Don't get me wrong, I loved the design you came up with to make SecretCredInjector reusable for different Secret Managers, but...it might be a tolerable sacrifice to get around the classloading issue.
3. We could just accept the current setup and require Solr admin's to manually copy around jars to use this component.

Anyone have thoughts?

Reply:
Thank you @gerlowskija for taking time to dig into this.
  
The issue here is that ZK connection to load _solr.xml_ is [called](https://github.com/apache/solr/blob/26b29d4df41e4efe67576efd72a603818c42738e/solr/core/src/java/org/apache/solr/core/NodeConfig.java#L213) before the modules are loaded.

The modules are loaded only after the above mentioned method returns a _NodeConfig_ whose constructor [calls](https://github.com/apache/solr/blob/26b29d4df41e4efe67576efd72a603818c42738e/solr/core/src/java/org/apache/solr/core/NodeConfig.java#L194) _initiModules()_.

I am trying to find a clean way to load the modules before _SolrZkClient_ is instantiated.

Edit: Actually there is no easy way to do this without an ugly refactoring of tons of classes. Not sure if it's really worth it.

A class loader is needed to load the modules. Now, _SolrZkClient_ doesn't have access to any core functionality, and there is no way to get a _loader_ from there.
One (ugly) solution would be to add a _SolrClassLoader_ to those _SolrZkClient_ constructors, but doing so you will need to inject it from all the calling classes (dozens), and it's not guaranteed all those classes do have a a _SolrClassLoader_ ...

Another solution would be to add a local custom static _SolrClassLoader_ class into _SolrZkClient_, but you don't have a _solrHome_...

There is no really good solution, maybe we just need to keep it simple (for now) and ask the admin to copy the libs, in same way it used to be for some pre-9 contribs (before this new modules loader feature has beed added)?

On the other hand, if you think about it, all the hassle here makes sense because Solrj has always been maintained as a special kind of module, with client (agnostic?) side compatibility in mind, not linked to the rest of Solr modules.

Reply:
Okay I ended up creating a new custom classLoader (_ModulesClassLoader_) and injected it into _SecretCredentialInjector_ and looks like it's working.
Will do some code clean up and more tests and push it upstream.

Reply:
> The issue here is that ZK connection to load solr.xml is called before the modules are loaded

True, if solr.xml lives in ZooKeeper. But it can also live on-disk on each Solr node.  When solr.xml lives on-disk, the CNFE shows up later, on CoreContainer loading (after NodeConfig has been fetched/parsed):

```
  at org.apache.solr.common.cloud.SolrZkClient.<init>(SolrZkClient.java:206)
org.apache.solr.common.SolrException: java.lang.IllegalArgumentException: Could not instantiate the SecretCredentialProvider class with name: org.apache.solr.secret.zk.AWSSecretCredentialsProvider
  at org.apache.solr.common.cloud.SolrZkClient.<init>(SolrZkClient.java:206) ~[?:?]
  at org.apache.solr.cloud.ZkController.<init>(ZkController.java:490) ~[?:?]
  at org.apache.solr.core.ZkContainer.initZooKeeper(ZkContainer.java:142) ~[?:?]
  at org.apache.solr.core.CoreContainer.load(CoreContainer.java:742) ~[?:?]
  at org.apache.solr.servlet.CoreContainerProvider.createCoreContainer(CoreContainerProvider.java:376) ~[?:?]
```

I wasn't terribly concerned with the case where the CNFE occurs trying to load solr.xml from ZK.  Primarily because the problem pre-exists your PR: users will hit that today in released versions of Solr if they bring their own ZkACLProvider implementation etc.   And secondarily because IMO it seems reasonable to say that if you want to customize how Solr connects to ZK, you need to put your configuration for that on-disk so that Solr can read it before ZK-conn time.

I'd sleep just fine at night if we just "punted" on that problem and documented the limitation.

That said, if you feel strongly that you want to solve the problem for all cases (i.e. where solr.xml lives in ZK), we can try that.  But Solr folks (not me unfortunately, but others) have put a lot of thought into Solr's classloader setup to support the many ways that Solr supports loading "extra" jars (contribs/modules, "plugins", blob store entries, etc.). That's not to say that we _can't_ introduce a new classloader to solve our problem here, but it'll take some very careful testing and probably some review by someone more familiar with Solr's classloader code than I am.

Reply:
Actually, it’s not only about solr.xml, _ZKController_ also needs also to create a zkClient, and the call is made in SolrZKClient (_solrj_) which doesn’t have access to _solr core_ class loaders.
Lot (all?) of communications with ZK are made through _solrj_. You were right earlier when you were saying that _SecretCredentialInjector_  is using the default classLoader. That means loading the modules doesn't change anything as long as you’re still using the default classLoader…

So a class loader  in _solrj_ is needed. Either we pass it from _solr core_ to _solrj_ through solrZkClient constructors (ugly) and by doing so we are creating a dependency between solrj and solr core, which I think goes against the philosophy of _solrj_.
Or
we create a custom _solrJ_ class loader completely independent from _core_. This is what I did with  _ModuleClassLoader_ 
 

On the other hand, I would rather ask the admin to copy few jars to some dir before starting than having them manage solr.xml with needs some expertise, not to mention the need to maintain it from different places, instead of having it in one place (zookeeper)…

I see only 2 options left:
1- Copy  the jars to _solr-webapp/webapp/WEB-INF/lib/_
2- Use custom class loader (which maybe useful for future features related to solrj including custom credential injectors).

I pushed a new commit with a solution based on a custom classLoader.

PS: The existing (new feature) to pass modules through  environment variable SOLR_MODULES doesn't work, only _-Dsolr.modules=scripting_ option works.

Reply:
> The existing (new feature) to pass modules through environment variable SOLR_MODULES doesn't work, only -Dsolr.modules=scripting option works.

Hmm, it works for me.  Not sure what's up there, but that's a discussion for a different JIRA and a different PR I guess...

> ZKController also needs to create a zkClient, and the call is made in SolrZKClient (solrj) which doesn’t have access to solr core class loaders.

True enough - but that default classloader usage is conditional - SolrZkClient only uses it when [the connection-strategy object doesn't already have a ZkCredentialsProvider](https://github.com/apache/solr/blob/main/solr/solrj/src/java/org/apache/solr/common/cloud/SolrZkClient.java#L155).  AFAICT from reading the code and some manual tests, the default classloader block will never run on the Solr server because Solr [always sets a ZkCredentialsProvider](https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L340) in ZkController.

That's why I suggested biting the bullet and creating a ZkCredentialsProvider impl that's AWS-specific, rather than creating the extra layer of indirection that runs afoul of the classloader mixing. Solr-core already has logic to ensure that ZkCredentialsProvider impl's don't hit classloader issues via its use of ZkClientConnectionStrategy.

We can get that result with the additional classloader solution you proposed too, but it makes me leery from a "principle of least change" perspective.  Adding a new classloader to Solr and SolrJ is a big change: conceptually if not in terms of LOC.

Reply:
I have a branch [here](https://github.com/gerlowskija/solr/commit/7468661cf94702b1dbc6386604de044b6bf26e70) that implements the "make an AWS-specific ZkCredentialsInjector" idea.

Like I said, I think this approach should be "safe" classpath wise and would be a smaller change conceptually for Solr to take on.  But I could still be missing something for sure and you've no reason to take my word for it.  If you're still not sure, maybe we could kick the tires a bit on that branch and see what we find?

Reply:
> True enough - but that default classloader usage is conditional - SolrZkClient only uses it when [the connection-strategy object doesn't already have a ZkCredentialsProvider](https://github.com/apache/solr/blob/main/solr/solrj/src/java/org/apache/solr/common/cloud/SolrZkClient.java#L155). AFAICT from reading the code and some manual tests, the default classloader block will never run on the Solr server because Solr [always sets a ZkCredentialsProvider](https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L340) in ZkController.
> 

You're right about about _ZKController_ but if you search for the usage of _SolrZkClient_ you will see there are lot of places where it's called without a connection strategy, in which case it creates the creds locally using the default classLoader.

Let me check that branch.

Reply:
Just went through this again, and again :) I like the idea of creating a ZkCredentialsInjector impl that's AWS-specific, rather than creating the extra layer of indirection (SecretCredentialsProvider) as @gerlowskija suggested.

This might pass tests, but would certainly need more tests and testing to check for places the completeness of this approach.

What I really appreciate about the approach here is that it does keep changes within a reasonable limit considering what we're trying to accomplish here.

I'll spend more time on this later this week and try testing this too.